### PR TITLE
DSBSparse: Addressing some performance issues

### LINKED ISSUE #4: Extend DBCSR get_block() to return sparse block
- Add optional argument to get_block(): dense : bool = False
- False by default to avoid memory allocation in the default case
- Return the requested block in the CSR format

### LINKED ISSUE #35: Adding missing tests for gpu_utils and mpi_utils
Need to add some tests in the gpu_utils and mpi_utils files. Look at codecov report to see the missing hits lines

### LINKED ISSUE #43: DSBSparse: Slow element-wise access
The current way of doing element-wise access of matrix elements can have some serious impact on performance, especially on GPU. It launches a gazillion of tiny, useless kernels.

- [ ] Rewrite element-wise access to be vectorized (slicing & fancy indexing).
- [ ] Think about how to write general GPU/CPU kernels for index resolution.

### LINKED ISSUE #46: Make unit test suite GPU/CPU agnostic
At the moment there are a lot of data transfers happening in the unit test suite that are unnecessary in a lot of cases. Of course this makes running the tests pretty slow on GPU...

This can be mostly fixed by just swapping `np.` for `xp.`.

### PR DESCRIPTION
**Pull request by vetschn**
Several things are still done naively in the DSBSparse matrices.

- [x] Caching mechanism for the DSBCOO block-wise access. Since we already have the rowptr map in the DSBCSR, there are no changes needed there. Also Mentioned in #4.
- [x] Return blocks in some sort of sparse format. Closes #4.
- [x] #43 
- [ ] ~~Better tests for arithmetic~~

### CODE REVIEW

**Comment by vetschn** [comment]
So i added these vectorized getters and setters now in the `DSBCOO` class (`DSBCSR` is next). The behavior is like this:

- If we are in the **`"stack"`** distribution state, you'll get an array of the expected shape (padded with zeros where requested elements are not in the matrix). So 
  ```python
  inds = np.arange(10)
  values = dsbcoo[inds, inds]
  ```
  will get `values.shape[-1] == 10` with maybe some zeros in there, where `(ind[i], ind[i])` is not in the matrix. You can also set items with the same logic:
  ```python
  inds = np.arange(10)
  values = np.ones(10)
  dsbcoo[inds, inds] = values
  ```
  This will only change those elements that are actually in the matrix, the sparsity pattern is immutable as always.

- If we are in the **`"nnz"`** distribution, the same logic should apply as before this change. If you request an element that is not in the matrix, you'll get an `IndexError` (because we cannot know which rank is supposed to hold an element that is not in the matrix). 

  If you request elements that are all in the matrix but distributed accross ranks, each rank now only returns the requested elements that it holds. On the ranks where there are no elements on the requested inds, you'll get an empty array (of the correct stack shape).
  ```python
  inds = np.arange(10)
  values = dsbcoo[inds, inds]
  ```
  could be `values.shape[-1] == 7 if comm.rank == 0`, `values.shape[-1] == 3 if comm.rank == 1`, and `values.shape[-1] == 0 if comm.rank == 2`.

  Setting elements in this distribution state works analogously:
    ```python
  inds = np.arange(10)
  values = np.ones(10)
  dsbcoo[inds, inds] = values
  ```
  Each rank only sets those elements of the input, which it owns. If we are accessing elements that are not in the matrix nothing happens and the sparsity remains unchanged.

Note that everything here also works when accessing only a part of the stack, so random things like
```python
inds = np.arange(10)
values = dsbcoo.stack[:3][inds, inds]
dsbcoo.stack[[2, 4, 6]][inds, inds] = 1.5 * values
```
should work as well in both distribution states.

File: src/qttools/__init__.py
Comment (vincent-maillou) [inline src/qttools/__init__.py:1]: I like that, didn't think about it but find it very clean.

File: src/qttools/utils/mpi_utils.py [hunk 1]
```diff
@@ -2,10 +2,12 @@
 
 from pathlib import Path
 
-import numpy as np
+import scipy.sparse as sps
```
Comment (vincent-maillou) [inline src/qttools/utils/mpi_utils.py:5]: Why do you still need scipy.sparse explicitly here? And don't use the sparse from your __init__?

Comment (vincent-maillou) [reply src/qttools/utils/mpi_utils.py:5]: EDIT: I guess it's to use the load_npz from host side, why not using: https://docs.cupy.dev/en/latest/reference/generated/cupy.load.html in the case of xp = cp?

File: tests/obc/test_spectral.py [hunk 2]
```diff
@@ -101,12 +100,12 @@ def test_memoizer(
     spectral = OBCMemoizer(spectral)
     a_ji, a_ii, a_ij = _make_periodic(a_xx, block_sections)
     x_ii = spectral(a_ii=a_ii, a_ij=a_ij, a_ji=a_ji, contact=contact)
-    assert np.allclose(x_ii, xp.linalg.inv(a_ii - a_ji @ x_ii @ a_ij), atol=1e-5)
+    assert xp.allclose(x_ii, xp.linalg.inv(a_ii - a_ji @ x_ii @ a_ij), atol=1e-5)
```
Comment (vincent-maillou) [inline tests/obc/test_spectral.py:103]: I'm not a huge fan of having the tolerances hard-coded here, I guess we  should have all solvers convergence tol somewhere? Anyway not a big deal

**Review by vincent-maillou — APPROVED** [review]
I'm not sure of the quality of my review as my brain stopped working a while ago today. But anyway: LGTM :1st_place_medal:

File: src/qttools/datastructures/dsbcoo.py [hunk 3]
```diff
@@ -147,29 +248,32 @@ def _get_block(self, stack_index: tuple, row: int, col: int) -> ArrayLike:
...
-            & (self.cols < self.block_offsets[col + 1])
-        )
-
         data_stack = self.data[*stack_index]
+        block_slice = self._get_block_slice(row, col)
+
+        if not self.return_dense:
+            if block_slice == slice(None):
+                # No data in this block, return an empty block.
+                return xp.empty(0), xp.empty(0), xp.empty(data_stack.shape[:-1] + (0,))
```
[the remark this answers is not in the document]
Comment (vetschn) [reply src/qttools/datastructures/dsbcoo.py:257]: This is the code that handles returning an empty "sparse" matrix, so empty `rows`, `cols` and `data`.

File: tests/obc/test_spectral.py (hunk 2 above)
Comment (vetschn) [reply tests/obc/test_spectral.py:103]: Definitely not a fan either. I think we talked about this already at some point but to test the Spectral OBC solver and the NEVP we should somehow construct matrices with hand-picked eigenvalues. That could maybe also solve this?

Alternatively, cranking up the refinement iterations always helps :sweat_smile:

File: src/qttools/datastructures/dsbcsr.py [hunk 4]
```diff
@@ -55,98 +54,175 @@ def __init__(
...
-            # that is not in the matrix, so we raise an error.
-            raise IndexError("Requested element not in matrix.")
+            mask = (brows == brow) & (bcols == bcol)
+            mask_inds = xp.where(mask)[0]
+
+            # Renormalize the row indices for this block.
+            rr = rows[mask] - self.block_offsets[brow]
+            cc = cols[mask]
+
+            # TODO: This could perhaps be done in an efficient way.
```
[the remark this answers is not in the document]
Comment (vetschn) [reply src/qttools/datastructures/dsbcsr.py:100]: See #50

File: src/qttools/datastructures/dsbcoo.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (vincent-maillou) [reply src/qttools/datastructures/dsbcoo.py:257]: make sense, thanks

File: tests/obc/test_spectral.py (hunk 2 above)
Comment (vincent-maillou) [reply tests/obc/test_spectral.py:103]: Yes, maybe open an issue for the eigenvalues solvers tests, I foresee @almaeder being very suited for this issue :smile:

File: src/qttools/datastructures/dsbcsr.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (vincent-maillou) [reply src/qttools/datastructures/dsbcsr.py:100]: amazing! thks

File: tests/obc/test_spectral.py (hunk 2 above)
Comment (vetschn) [reply tests/obc/test_spectral.py:103]: See #51

**Pull request merged**